rfcs/2021-08-06-8619-framing-and-codecs-sources.md

```diff
@@ -0,0 +1,212 @@
+parser to create an `Event` and returns.
+
+Sources which want to expose framing/decoding functionality to the user can
+embed `DecodingConfig` in their config, build the `Decoder` and apply it to the
+sequences of bytes they produce to create events.
+
+## Doc-level Proposal
+
+The `framing` and `decoding` behavior can be configured for a source in the
+vector config, e.g.:
+
+```yaml
+[framing]
+type = "character_delimited"
+delimiter = "\t"
+max_length = 1337
+
+[decoding]
+type = "json"
+```
+
+## Rationale
+
+One often request feature is reading JSON-encoded messages from Kafka.
+Currently, this can only be accomplished by configuring and wiring together two
+separate components, where with codecs it could be a convenient one-line change
+to the config.
+
+Introducing framing/decoding, a source's implementation may also be reused
+internally. One example would be the `syslog` source (see
+[#7046](https://github.com/timberio/vector/pull/7046)), or the upcoming `syslog`
+sink in [#7106](https://github.com/timberio/vector/issues/7106). Instead of
+re-implementing socket-based connection handling, the `syslog` components could
+be replaced by the `socket` counterparts combined with `octet-framing`. This
+reduces a possible source of bugs and inconsistencies and therefore leads to
+less maintenance burden.
+
+Introducing codecs may also shrink unnecessary noise in config files by removing
+transform steps / input indirections, when basic transforms were used that are
+only concerned with encoding formats.
+
+## Prior Art
+
+[Tokio Codecs](https://docs.rs/tokio-util/0.6.7/tokio_util/codec/index.html)
+provide traits to conveniently convert from `AsyncRead`/`AsyncWrite` to
+`Stream`/`Sink`. These are currently used in custom implementations of sources
+to frame byte streams. We want to extend this existing facility to produce
+events from bytes directly.
+
+[Logstash Codec Plugins](https://www.elastic.co/guide/en/logstash/current/codec-plugins.html)
+are interesting since they operate on a higher level than what has been proposed
+in this RFC. They don't distinguish between a framing and codec stage, but
+rather have codecs that support framing (e.g. `line` codec), compression (e.g.
+`gzip_lines` codec) and encoding (e.g. `protobuf` codec). Supporting these kind
+of codecs could be an interesting future thought but would require an
+architectural change, especially to the internal representation of an `Event`.
+
+## Drawbacks
+
+It is possible that the proposed abstraction is too rigid, in a sense that it is
+not possible to cleanly separate these stages into "framing" and "codec"
+responsibilities. This limits what codecs can do, e.g. applying framing after
+decoding/decompression would not be possible as there are no means to "go back"
+to the framing stage. However, the proposed solution is still strictly better
+than the status quo, as it provides a consistent interface.
```

Comment:
Might also be worth noting that the proposal here doesn't conflict with doing something more general in the future. We could wrap these decoders in a more general transform-like structure if there's strong demand for that, and we'd have the benefit of these simpler building blocks.

Reply:
Added in a3caac03324016a4e3334ea865c1bc2a7e29f1cd